### PR TITLE
grok to understand ARG directive and build args

### PR DESCRIPTION
**Pull request by shhsu**
**Purpose of the PR:**

**Fixes #**
https://github.com/Azure/acr-builder/issues/83

**Notes/Details:**

dependency scanning now knows build args

```release-note
```

@Azure/azure-container-registry

### CODE REVIEW

File: pkg/grok/docker_file.go [hunk 1]
```diff
@@ -10,41 +10,66 @@ import (
...
+			case "FROM":
+				if len(tokens) < 2 {
 					return "", nil, fmt.Errorf("Unable to understand line %s", line)
 				}
-				alias := tokens[3]
-				originLookup[alias] = origin
-				// Just ignore the rest of the tokens...
-				if len(tokens) > 4 {
-					logrus.Debugf("Ignoring chunks from FROM clause: %v", tokens[4:])
+				var image = os.Expand(tokens[1], func(key string) string {
```
Comment (northtyphoon) [inline pkg/grok/docker_file.go]: Cool it looks os.Expand will handle $arg, ${arg} very well?

Comment (shhsu) [reply pkg/grok/docker_file.go]: yes, I used it in build.Context too. Just forgot about it at the moment during our offline conversation.

As you can see both $arg and ${arg} are used in our test example.

**Comment by ehotinger** [comment]
btw @northtyphoon @shhsu didn't have a chance to review this, but does it work for `ENV={blah}` too? There's docs on Docker where all `{}` can be used so just making sure we covered them all 👍

**Comment by shhsu** [comment]
As far as I can tell, you cannot use ENV variable on FROM clause.
ARG goes before FROM and ENV goes after.
I tried a few multi-stage example and none of them seem to work. @ehotinger, do you have any example for ENV applying for this use case?

**Comment by ehotinger** [comment]
@shhsu ya original syntax I posted is a bit off, basically just want to make sure this still works after replacement:

```
ENV abc=hello
ENV abc=bye def=$abc
ENV ghi=$abc
```

I'll review the PR later when I have some free time.

**Comment by shhsu** [comment]
Actually, this PR only apply to dependency scanning logic. If a variable is not used in FROM clause we don't really care about it.

**Pull request merged**